### PR TITLE
feat!: get module info from `modules.json`

### PR DESCRIPTION
**Pull request by matt-FFFFFF**
If `module_path` is set, it will examine the `modules.json` file and populate the calculated attributes of: `module_source` and `module_version`.
 
The PR supports the standard `.terraform` directory as well as those who have customised with `$TF_DATA_DIR`.
 
The code supports using this with a standard module, or a submodule. We should have telemetry resources in the submodule too for this to work.
 
Tests added for the above scenarios.

Removed support for the nonce and ephemeral number

> Note: gosec failing as we are using a higher go version that it supports

### CODE REVIEW

File: internal/provider/telemetry_resource.go [hunk 1]
```diff
@@ -140,14 +145,23 @@ func (r *TelemetryResource) Create(ctx context.Context, req resource.CreateReque
 		return
 	}
 
+	data.ModuleSource = basetypes.NewStringNull()
```
Comment (jaredfholgate) [inline internal/provider/telemetry_resource.go:148]: As above, can we fallback to checking for a supplied source and version if the module_path is not supplied.

File: internal/provider/telemetry_resource.go [hunk 2]
```diff
@@ -160,7 +174,22 @@ func (r *TelemetryResource) Read(ctx context.Context, req resource.ReadRequest,
 	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
 
 	if resp.Diagnostics.HasError() {
-		return
+		resp.Diagnostics.Append()
+	}
+
+	data.ModuleSource = basetypes.NewStringNull()
```
Comment (jaredfholgate) [inline internal/provider/telemetry_resource.go:180]: Duplicate code. Would it make sense to extract a function for this?

**Review by jaredfholgate — APPROVED** [review]
LGTM, just some suggestions that can be ignored.

File: internal/provider/telemetry_resource.go (hunk 2 above)
Comment (matt-FFFFFF) [reply internal/provider/telemetry_resource.go:180]: Good idea - added `data = updateModuleSourceAndVersion(data)`

File: internal/provider/telemetry_resource.go (hunk 1 above)
Comment (matt-FFFFFF) [reply internal/provider/telemetry_resource.go]: Resolved in 6582745

File: internal/provider/map_validator.go [hunk 3]
```diff
@@ -18,9 +24,11 @@ func (m mapValidator) MarkdownDescription(ctx context.Context) string {
 }
 
 func (m mapValidator) ValidateMap(ctx context.Context, request validator.MapRequest, response *validator.MapResponse) {
+	reservedKeys := []string{"event", "source", "version"}
```
[the remark this answers is not in the document]
Comment (matt-FFFFFF) [reply internal/provider/map_validator.go]: thanks! fixed in 6582745

**Comment by matt-FFFFFF** [comment]
@lonegunmanb please can you review

File: internal/provider/telemetry_resource_test.go [hunk 4]
```diff
@@ -931,12 +921,187 @@ resource "modtm_telemetry" "test" {
...
+      "Dir": ".terraform/modules/keys/modules/key"
+    },
+    {
+      "Key": "kv",
+      "Source": "registry.terraform.io/Azure/avm-res-keyvault-vault/azurerm",
+      "Version": "0.6.1",
+      "Dir": ".terraform/modules/kv"
+    },
+    {
+      "Key": "kv.keys",
```
Comment (lonegunmanb) [inline internal/provider/telemetry_resource_test.go:1046]: Should we add test case for this item? Could our current implementation handle this module correctly?

File: internal/provider/telemetry_resource_test.go [hunk 5]
```diff
@@ -931,12 +921,187 @@ resource "modtm_telemetry" "test" {
...
+      "Dir": ".terraform/modules/kv/modules/secret"
+    }
+  ]
+}
+`)
+	return err
+}
+
+func TestUpdateModuleSourceAndVersion(t *testing.T) {
+	require.NoError(t, createModulesJson())
```
Comment (lonegunmanb) [inline internal/provider/telemetry_resource_test.go:1062]: Skip this test on Windows.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Hi @matt-FFFFFF thanks for opening this pr, a few comments.

File: internal/provider/telemetry_resource_test.go [hunk 6]
```diff
@@ -931,12 +921,187 @@ resource "modtm_telemetry" "test" {
...
+			_, err := modulePathToKey(tt.modulePath)
+			assert.Error(t, err)
+		})
+	}
+}
+
+func TestModulePathToKey_WithCustomTFDataDir(t *testing.T) {
+	customTFDataDir := "/custom/terraform/data"
+	t.Setenv("TF_DATA_DIR", customTFDataDir)
+	defer os.Unsetenv("TF_DATA_DIR")
```
Comment (lonegunmanb) [inline internal/provider/telemetry_resource_test.go:1000]: Based on [document](https://beta.pkg.go.dev/testing@master#T.Setenv), I think we can skip `Unsetenv` manually here.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]

**Comment by lonegunmanb** [comment]
Another thought is, could we convert `Dir` in `modules.json` and module path we received by using `filepath.Abs` function, then match the key?

**Review by lonegunmanb — APPROVED** [review]
LGTM

**Pull request merged**